Drawer: allow a custom element to be focused on close (#1669)
agriculturegovau/agds-next: packages/react/src/drawer/docs/overview.mdx

	);
};
```

## Focus custom element on close

> When the drawer closes, focus is returned to the element that triggered its opening. You can override this behaviour by passing a custom element to the `elementToFocusOnClose` prop. This is useful when an action performed in the drawer causes an error and focus should be returned to an alert on close.

Reviewer: OK how about this?
```suggestion
When a drawer closes, focus is returned to the element that opened it. You can override this behaviour, by passing an element ref to the `elementToFocusOnClose` prop. This can be used in situations such as when an action performed in the drawer causes an error on the main page, thus requiring the user’s focus to be sent to the alert instead.
```

```jsx live
() => {
	const sectionAlertRef = React.useRef(null);
	const [error, setError] = React.useState('');
	const [isDrawerOpen, openDrawer, closeDrawer] = useTernaryState(false);

	const openDrawerAndClearError = () => {
		openDrawer();
		setError('');
	};

	const closeWithError = () => {
		closeDrawer();
		setError('We were unable to perform the action, please try again later.');
	};

	return (
		<Stack gap={2}>
			<SectionAlert

> 				focusAll

Reviewer: Redundant

```suggestion
```

	);
};
```

## Focus custom element on close

> When a drawer closes, focus is returned to the element that opened it. You can override this behaviour by passing an element to the `elementToFocusOnClose` prop. This can be used in situations, for example, when an action performed in the drawer causes an error, setting the main page in an error state, and requiring the user’s focus to be returned to the alert instead.

Author: @stowball I've updated this based on our conversation. Let me know what you think.

Reviewer: Looks good to me!

	);
};
```

## Focus custom element on close

> When the drawer closes, focus is returned to the element that triggered its opening. You can override this behaviour by passing a custom element to the `elementToFocusOnClose` prop. This is useful when an action performed in the drawer causes an error and focus should be returned to an alert on close.

Reviewer: Perhaps something like this?
```suggestion
By default, when a drawer closes, focus is returned to the element that opened it. Should you need to override this behaviour, pass a element ref to the `elementToFocusOnClose` prop. This can be used in situations such as when an action performed in the drawer causes an error on the main page, thus requiring the user’s focus to be sent to the alert instead.
```

Author: I tried to spice it up by not re-using the "By default" opener that already exists on this docs page.

```jsx live
() => {
	const sectionAlertContainerRef = React.useRef(null);
	const [error, setError] = React.useState('');
	const [isDrawerOpen, openDrawer, closeDrawer] = useTernaryState(false);

	const openDrawerAndClearError = () => {
		openDrawer();
		setError('');
	};

	const closeWithError = () => {
		closeDrawer();
		setError('We were unable to perform the action, please try again later.');
	};

	return (
		<Stack gap={2}>
			<Box
				css={error ? undefined : { display: 'none' }}

> 				focusRingFor="all"
> 				ref={sectionAlertContainerRef}
> 				tabIndex={-1}

Reviewer: None of this should exist?
```suggestion
```

Author: Yeah, all this is needed. It's not needed on the alert, though. Drawer is pernickety.

Reviewer: Isn't it better to have it on the alert instead of this outer element? It seems to still function

Author: Hmm it didn't in my testing. Did you edit the docs or the code?

There's no `ref.current` when the drawer opens if you conditionally render the alert. `ref` updates don't trigger component re-renders so the `elementToFocusOnClose` prop never updates and no focus is set via the drawer.

Reviewer: It should be this:

```
<Box css={error ? undefined : { display: 'none' }}>
	<SectionAlert
		title="Something went wrong"
		tone="error" focusRingFor="all"
		ref={sectionAlertContainerRef}
		tabIndex={-1}
	>
		<Text as="p">{error}</Text>
	</SectionAlert>
</Box>
```
			
Otherwise no focus ring will be applied since `tabIndex={-1}` causes the focus ring for all on `SectionAlert`, but not for `Box`

Reviewer: Actually, your solution should work too, but doesn't…? Why isn't the focus ring being applied? I'd still prefer my solution since we should encourage focusing the actual alert, but we need to figure out why yours doesn't do anything

			>
				<SectionAlert

> 					isHidden={Boolean(!error)}

Reviewer: ```suggestion
```
